dgl_sparse/src/softmax.cc

```diff
@@ -0,0 +1,97 @@
+/**
+ *  Copyright (c) 2022 by Contributors
+ * @file softmax.cc
+ * @brief DGL C++ Softmax operator implementation
+ */
+
+#include <sparse/sparse_matrix.h>
+#include <torch/script.h>
+
+#include "./matmul.h"
+#include "./utils.h"
+
+namespace dgl {
+namespace sparse {
+
+using namespace torch::autograd;
+
+class SoftmaxAutoGrad : public Function<SoftmaxAutoGrad> {
+ public:
+  static torch::Tensor forward(
+      AutogradContext* ctx, c10::intrusive_ptr<SparseMatrix> sparse_mat,
+      torch::Tensor sparse_val);
+
+  static tensor_list backward(AutogradContext* ctx, tensor_list grad_outputs);
+};
+
+void _SoftmaxSanityCheck(
+    c10::intrusive_ptr<SparseMatrix> sparse_mat, torch::Tensor sparse_val) {
+  auto val_shape = sparse_val.sizes();
+  CHECK_EQ(val_shape[0], sparse_mat->nnz())
+      << "Softmax: the value shape does not match nnz of the sparse matrix.";
+  CHECK(sparse_val.device() == sparse_mat->device())
+      << "Softmax: sparse matrix and non-zero values should be on the same "
+         "device.";
+}
+
+torch::Tensor SoftmaxAutoGrad::forward(
+    AutogradContext* ctx, c10::intrusive_ptr<SparseMatrix> sparse_mat,
+    torch::Tensor sparse_val) {
+  auto sparse_val_max = SpMMNoAutoGrad(sparse_mat, sparse_val, "max");
```

Comment:
Instead of using SpMM, I feel the new sparse reduction operators are more suitable. https://github.com/dmlc/dgl/blob/master/dgl_sparse/include/sparse/reduction.h

Reply:
replaced